File: most-frequent-subtree-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findFrequentTreeSum(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []

        frequencies = {}

        def dfs(node):
            if not node:
                return 0
            sumLeft = dfs(node.left)
            sumRight = dfs(node.right)
            sumAll = sumLeft + sumRight + node.val
            if sumAll in frequencies:
                frequencies[sumAll] += 1
            else:
                frequencies[sumAll] = 1
            return sumAll

        dfs(root)

        maxFrequency = max(frequencies.values())
        return [key for key in frequencies.keys() if maxFrequency == frequencies[key]]

File: most-frequent-subtree-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findFrequentTreeSum = function(root) {
  if (!root) {
    return []
  }

  const frequencies = {}

  function inOrder (node) {
    if (!node) {
      return 0
    }
    const sum = inOrder(node.left) + node.val + inOrder(node.right)
    frequencies[sum] = sum in frequencies ? frequencies[sum] + 1 : 1
    return sum
  }

  inOrder(root)

  const maxFrequency = Math.max.apply(null, Object.values(frequencies))

  const sums = []
  for (let key in frequencies) {
    if (frequencies[key] === maxFrequency) {
      sums.push(parseInt(key))
    }
  }
  return sums
};
